src/azure/azureResourceController.ts

```diff
@@ -0,0 +1,68 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Location } from '@azure/arm-subscriptions';
+import { ResourceGroup } from '@azure/arm-resources';
+import { Server } from '@azure/arm-sql';
+import * as mssql from 'vscode-mssql';
+import * as azureUtils from './utils';
+
+export class AzureResourceController {
+
+	constructor(
+		private _subscriptionClientFactory: azureUtils.SubscriptionClientFactory = azureUtils.defaultSubscriptionClientFactory,
+		private _resourceManagementClientFactory: azureUtils.ResourceManagementClientFactory = azureUtils.defaultResourceManagementClientFactory,
+		private _sqlManagementClientFactory: azureUtils.SqlManagementClientFactory = azureUtils.defaultSqlManagementClientFactory) {
+	}
+
+	/**
+	 * Returns Azure locations for given session
+	 */
+	public async getLocations(session: mssql.IAzureAccountSession): Promise<Location[]> {
+		const subClient = this._subscriptionClientFactory(session.token);
+		if (session.subscription?.subscriptionId) {
+			const locationsPages = await subClient.subscriptions.listLocations(session.subscription.subscriptionId);
+			let locations = await azureUtils.getAllValues(locationsPages, (v) => v);
+			return locations.sort((a, b) => (a.name || '').localeCompare(b.name || ''));
+		} else {
+			throw new Error('Invalid session');
```

Comment:
localize

Reply:
I noticed that we didn't localize the error coming from extension API. I thought it might be intentional, so I did the same